### PR TITLE
tweaking nn parameters 

### PR DESCRIPTION
**Pull request by wiederm**
## Description
sets up the infrastructure to tweak the nn parameters trying to minimize the loss defined as the difference between experimental (in solution) and calculated (in vacuum) free energy difference. Currently set up for a single molecule. 

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [ ] expend from one tautomer pair to all tautomer pairs
  -[ ] quality control of the changed neural net parameters

## Questions
- [ ] how to best approach the quality control aspect

### CODE REVIEW

**Comment by maxentile** [comment]
Hah, I hadn't noticed Github ever give up on counting the changed files before!
image

I'll plan to review carefully on Monday. Specifically I'd like to review the aspects related to batching, but probably not most of the other changes in this PR.

Should I mostly be looking at the file parameter_gradients.py, or are there other places I should be reviewing also?

**Comment by wiederm** [comment]
I think the parameter_gradients.py is good place to start! I have removed large chunks of the code that were not used by our current vacuum approach. I think it made the code much harder to understand. We can always just roll back if needed.

The batching update was mostly in the ani.py and restrain.py part -- if you have time reviewing these changes would also be good. These are covered by tests, but a second pair of eyes is always good!

One thing to note -- and which will become apparent during the review -- is the inconsistency between [floats * unit, ...] vs [floats, ...] * unit. it's super annoying to convert these (and some functions need a specific form) but I have not figured out a consistent way to treat this.

**Comment by maxentile** [comment]
Nice, thanks for the orienting comments and heads-up! Will focus there.

**Comment by maxentile** [comment]
Looks great!

Unfortunately I wasn't able to add comments to specific lines, because Github can't display a diff this big:

image

Minor per-file comments, non-blocking:
* `scripts/parameter_gradient-opt.py` -- I would include in the comments a reference to [the torchani example code](https://aiqm.github.io/torchani/examples/nnp_training.html) some of this is based on. I would also comment on any important differences relative to that example (e.g. `weight_decay` parameters look different -- is that important? `scheduler.step(rmse)` --> `scheduler.step(rmse/10)` -- is that important?)
* `parameter_gradients.py` -- in `tweak_parameters(names, data_path, nr_of_nn, max_epochs)` function, I have the same comments as above: would add a link to example code, and comment on differences relative to it. I also note that this function itself contains function definitions, which could possibly be popped out and reused. (For example, to reduce duplication in the `parameter_gradient-opt.py` script.) I would ideally prefer to have (1) a function that expresses a free energy estimate as a differentiable function of model parameters, (2) a function that compares free energy estimates to arbitrary target values (`rmse`), (3) a function to optimize torchani-specific model parameters to make a loss-function (such as `rmse` -- or later `rmse` with some penalties) small -- and incorporating any torchani-specific Adam/SGD scheduling tricks (and exposing torchani-specific options, such as to fine-tune just the last layer's parameters vs. to all parameters to vary), (4) a function that reads experimental free energy values, sampled configuration snapshots, etc. from a specified file-path to set up the problem (the part of `tweak_parameters` that looks at `names`, `data_path`). Currently all 4 of these goals are a bit entangled in the `tweak_parameters` function.
